internal/cluster/cluster.go

```diff
@@ -0,0 +1,264 @@
+			logger.Error(err, "sign failed: %v", err)
+			continue
+		}
+
+		if buildRequeue || signRequeue {
+			requeue = true
+		}
+	}
+
+	return requeue
+}
+
+func (c *clusterAPI) GarbageCollectBuilds(ctx context.Context, mcm kmmv1beta1.ManagedClusterModule) ([]string, error) {
+	namespace := c.defaultJobNamespace
+	if mcm.Spec.JobNamespace != "" {
+		namespace = mcm.Spec.JobNamespace
+	}
+	return c.buildAPI.GarbageCollect(ctx, mcm.Name, namespace, &mcm)
+}
+
+func (c *clusterAPI) kernelMappingsByKernelVersion(
+	ctx context.Context,
+	modKernelMappings []kmmv1beta1.KernelMapping,
+	cluster clusterv1.ManagedCluster) (map[string]*kmmv1beta1.KernelMapping, error) {
+
+	kernelVersions, err := c.kernelVersions(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	mappings := make(map[string]*kmmv1beta1.KernelMapping)
+	logger := log.FromContext(ctx)
+
+	for _, kernelVersion := range kernelVersions {
+		osConfig := c.kernelAPI.GetNodeOSConfigFromKernelVersion(kernelVersion)
+		kernelVersion := strings.TrimSuffix(kernelVersion, "+")
+
+		kernelVersionLogger := logger.WithValues(
+			"kernel version", kernelVersion,
+		)
+
+		if image, ok := mappings[kernelVersion]; ok {
+			kernelVersionLogger.V(1).Info("Using cached image", "image", image)
+			continue
+		}
+
+		m, err := c.kernelAPI.FindMappingForKernel(modKernelMappings, kernelVersion)
+		if err != nil {
+			kernelVersionLogger.Info("no suitable container image found; skipping kernel version")
+			continue
+		}
+
+		m, err = c.kernelAPI.PrepareKernelMapping(m, osConfig)
+		if err != nil {
+			kernelVersionLogger.Info("failed to substitute the template variables in the mapping", "error", err)
+			continue
+		}
+
+		kernelVersionLogger.V(1).Info("Found a valid mapping",
+			"image", m.ContainerImage,
+			"build", m.Build != nil,
+		)
+
+		mappings[kernelVersion] = m
+	}
+
+	return mappings, nil
+}
+
+func (c *clusterAPI) kernelVersions(cluster clusterv1.ManagedCluster) ([]string, error) {
+	for _, clusterClaim := range cluster.Status.ClusterClaims {
+		if clusterClaim.Name != clusterClaimName {
+			continue
+		}
+		return strings.Split(clusterClaim.Value, "\n"), nil
+	}
+	return nil, errors.New("KMM ClusterClaim not found")
+}
+
+func (c *clusterAPI) build(
+	ctx context.Context,
+	mod kmmv1beta1.Module,
+	mcm *kmmv1beta1.ManagedClusterModule,
+	kernelMapping *kmmv1beta1.KernelMapping,
+	kernelVersion string) (bool, error) {
+
+	shouldSync, err := c.buildAPI.ShouldSync(ctx, mod, *kernelMapping)
+	if err != nil {
+		return false, fmt.Errorf("could not check if build synchronization is needed: %v", err)
+	}
+	if !shouldSync {
+		return false, nil
+	}
+
+	logger := log.FromContext(ctx).WithValues(
+		"kernel version", kernelVersion,
+		"image", kernelMapping.ContainerImage)
+	buildCtx := log.IntoContext(ctx, logger)
+
+	buildRes, err := c.buildAPI.Sync(buildCtx, mod, *kernelMapping, kernelVersion, true, mcm)
+	if err != nil {
+		return false, fmt.Errorf("could not synchronize the build: %w", err)
+	}
```

Comment:
This logic is also used in `controllers/module_reconciler.go`. Can we extract that functionality to a separate API (utils for example) and reuse it instead of duplicating it?

Reply:
Good point and I was also thinking about this when implementing this part. I was trying to find a way to refactor the `buildAPI` itself to accommodate for this combined functionality, i.e. a call that checks whether a sync should run or not and runs the actual `sync`. I realized that it would be better to skip adding a layer on top of the `buildAPI` in another internal package.

So while I also thought that this is duplicating code, I figured I'd leave as is, because I couldn't find a better trade-off. 

WDYT? Should we go about adding a layer or refactor the current `buildAPI` (or maybe a 3rd option that I'm not thinking about)? 

PS. if we decide to refactor the current `buildAPI`, I think that we could do that in a separate PR. So that we do it with more focus and to not make this one bigger.

Reply:
@ybettan @yevgeny-shnaidman @qbarrand maybe there are more ideas, as to how to tackle this?

Reply:
We will probably schedule a refactoring of the code after GA, we have a few places where we can improve our code, so lets get back to that issue then

Reply:
Duly noted, I agree with @yevgeny-shnaidman and I believe we can afford to come back to this piece of code after GA.